File: packages/js/openvino/src/xml_builder.ts
/* eslint-disable */
/**
 * Represents an XML node for building OpenVINO IR files.
 */
export class XmlNode {
  /** The name of the XML tag. */
  name: string;
  /** Attributes of the XML tag. */
  attributes: Record<string, string>;
  /** Children of the XML tag, can be other XmlNodes or strings. */
  children: (XmlNode | string)[];

  /**
   * Creates a new XmlNode.
   * @param name - The name of the XML tag.
   * @param attributes - Initial attributes.
   * @param children - Initial children.
   */
  constructor(
    name: string,
    attributes: Record<string, string> = {},
    children: (XmlNode | string)[] = [],
  ) {
    this.name = name;
    this.attributes = attributes;
    this.children = children;
  }

  /**
   * Sets an attribute on the XML node.
   * @param key - The attribute name.
   * @param value - The attribute value.
   * @returns The current XmlNode for chaining.
   */
  setAttribute(key: string, value: string): this {
    this.attributes[key] = value;
    return this;
  }

  /**
   * Adds a child to the XML node.
   * @param child - The child to add.
   * @returns The current XmlNode for chaining.
   */
  addChild(child: XmlNode | string): this {
    this.children.push(child);
    return this;
  }

  /**
   * Converts the XML node to a string.
   * @param indent - The current indentation level.
   * @param pretty - Whether to format the output with newlines and indentation.
   * @returns The XML string representation.
   */
  toString(indent: number = 0, pretty: boolean = false): string {
    const indentStr = pretty ? ' '.repeat(indent) : '';
    const newline = pretty ? '\n' : '';

    let attrStr = '';
    for (const [key, value] of Object.entries(this.attributes)) {
      const escapedValue = String(value)
        .replace(/&/g, '&amp;')
        .replace(/</g, '&lt;')
        .replace(/>/g, '&gt;')
        .replace(/"/g, '&quot;')
        .replace(/'/g, '&apos;');
      attrStr += ` ${key}="${escapedValue}"`;
    }

    if (this.children.length === 0) {
      return `${indentStr}<${this.name}${attrStr} />${newline}`;
    }

    if (this.children.length === 1 && typeof this.children[0] === 'string') {
      const escapedText = this.children[0]
        .replace(/&/g, '&amp;')
        .replace(/</g, '&lt;')
        .replace(/>/g, '&gt;');
      return `${indentStr}<${this.name}${attrStr}>${escapedText}</${this.name}>${newline}`;
    }

    let result = `${indentStr}<${this.name}${attrStr}>${newline}`;
    const childIndent = pretty ? indent + 4 : 0;

    for (const child of this.children) {
      if (typeof child === 'string') {
        const escapedText = child
          .replace(/&/g, '&amp;')
          .replace(/</g, '&lt;')
          .replace(/>/g, '&gt;');
        if (pretty) {
          result += ' '.repeat(childIndent) + escapedText + newline;
        } else {
          result += escapedText;
        }
      } else {
        result += child.toString(childIndent, pretty);
      }
    }

    result += `${indentStr}</${this.name}>${newline}`;
    return result;
  }
}

/**
 * A builder for creating XML documents.
 */
export class XmlBuilder {
  /** The root node of the XML document. */
  root: XmlNode | null = null;
  /** The XML declaration string. */
  declaration: string = '<?xml version="1.0" ?>';

  /**
   * Creates a new XmlBuilder.
   * @param rootName - Optional name for the root node.
   */
  constructor(rootName?: string) {
    if (rootName) {
      this.root = new XmlNode(rootName);
    }
  }

  /**
   * Sets the root node of the XML document.
   * @param node - The root node.
   * @returns The current XmlBuilder for chaining.
   */
  setRoot(node: XmlNode): this {
    this.root = node;
    return this;
  }

  /**
   * Sets the XML declaration string.
   * @param dec - The declaration string.
   * @returns The current XmlBuilder for chaining.
   */
  setDeclaration(dec: string): this {
    this.declaration = dec;
    return this;
  }

  /**
   * Converts the XML document to a string.
   * @param pretty - Whether to format the output with newlines and indentation.
   * @returns The XML document string representation.
   */
  toString(pretty: boolean = false): string {
    const newline = pretty ? '\n' : '';
    let result = this.declaration ? `${this.declaration}${newline}` : '';
    if (this.root) {
      result += this.root.toString(0, pretty);
    }
    return pretty ? result.trimEnd() : result;
  }
}

File: packages/python/onnx9000-openvino/src/onnx9000/openvino/xml_builder.py
"""Provides XML building utilities."""

from typing import Optional, Union


class XmlNode:
    """Represents an XML node."""

    def __init__(
        self,
        name: str,
        attributes: Optional[dict[str, str]] = None,
        children: Optional[list[Union["XmlNode", str]]] = None,
    ):
        """Initialize the XML node."""
        self.name = name
        self.attributes = attributes or {}
        self.children = children or []

    def set_attribute(self, key: str, value: str) -> "XmlNode":
        """Set an attribute on the node."""
        self.attributes[key] = str(value)
        return self

    def add_child(self, child: Union["XmlNode", str]) -> "XmlNode":
        """Add a child node or text."""
        self.children.append(child)
        return self

    def to_string(self, indent: int = 0, pretty: bool = False) -> str:
        """Convert the node to an XML string."""
        indent_str = " " * indent if pretty else ""
        newline = "\n" if pretty else ""

        attr_str = ""
        for key, value in self.attributes.items():
            escaped_value = (
                str(value)
                .replace("&", "&amp;")
                .replace("<", "&lt;")
                .replace(">", "&gt;")
                .replace('"', "&quot;")
                .replace("'", "&apos;")
            )
            attr_str += f' {key}="{escaped_value}"'

        if not self.children:
            return f"{indent_str}<{self.name}{attr_str} />{newline}"

        if len(self.children) == 1 and isinstance(self.children[0], str):
            escaped_text = (
                self.children[0].replace("&", "&amp;").replace("<", "&lt;").replace(">", "&gt;")
            )
            return f"{indent_str}<{self.name}{attr_str}>{escaped_text}</{self.name}>{newline}"

        result = f"{indent_str}<{self.name}{attr_str}>{newline}"
        child_indent = indent + 4 if pretty else 0

        for child in self.children:
            if isinstance(child, str):
                escaped_text = child.replace("&", "&amp;").replace("<", "&lt;").replace(">", "&gt;")
                result += f"{' ' * child_indent if pretty else ''}{escaped_text}{newline}"
            else:
                result += child.to_string(child_indent, pretty)

        result += f"{indent_str}</{self.name}>{newline}"
        return result


class XmlBuilder:
    """Builds an XML document."""

    def set_root(self, node):
        """Set the root node."""
        self.root = node
        return self

    def set_declaration(self, decl):
        """Set the XML declaration."""
        self.declaration = decl
        return self

    def __init__(self, root_name: Optional[str] = None):
        """Initialize the builder."""
        self.root: Optional[XmlNode] = XmlNode(root_name) if root_name else None
        self.declaration = '<?xml version="1.0" ?>'

    def to_string(self, pretty: bool = False) -> str:
        """Convert the document to a string."""
        newline = "\n" if pretty else ""
        result = f"{self.declaration}{newline}" if self.declaration else ""
        if self.root:
            result += self.root.to_string(0, pretty)
        return result.rstrip() if pretty else result
